packager/amp_cache_transform/amp_cache_transform.go

```diff
@@ -71,13 +80,102 @@ func parseParameterisedIdentifier(reader *strings.Reader) (*parameterisedIdentif
-	// NOTE: The initial version of AMP-Cache-Transform does not support
-	// any parameters, so we don't bother to implement a parser for them.
-	// No point in distinguishing syntactic failure from semantic failure.
-	// Once we add parameters, we'll need to implement parsers for the
-	// subset item that they use. See Items here:
-	// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-3.4
-	return &parameterisedIdentifier{primaryID}, nil
+	// NOTE: The current version of AMP-Cache-Transform only uses
+	// string-valued parameters. Future versions of the spec may require
+	// parsing other Item types.
+	params := map[string]string{}
+	for {
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+
+		// If we've reach the end of the parameter list, exit success.
+		if reader.Len() == 0 {
+			break
+		}
+		semicolon, err := reader.ReadByte()
+		if err != nil {
+			return nil, errors.Wrap(err, "reading ';'")
+		}
+		if semicolon != ';' {
+			if err := reader.UnreadByte(); err != nil {
+				return nil, errors.Wrap(err, "unreading ';'")
+			}
+			break
+		}
+
+		// Otherwise, parse a parameter name.
+		if err := discardOWS(reader); err != nil {
+			return nil, errors.Wrap(err, "discarding OWS")
+		}
+		name, err := parseIdentifier(reader)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing param name")
+		}
+		if _, has := params[name]; has {
+			return nil, errors.Errorf("param %q already present", name)
+		}
+		if name != "v" {
+			return nil, errors.Errorf("invalid AMP-Cache-Transform param %q", name)
+		}
+
+		// ... and a parameter value.
+		// NOTE: The current version of the AMP-Cache-Transform spec
+		// does not use any null-valued parameters. So, for now, a
+		// missing '=' is a parse failure.
+		equals, err := reader.ReadByte()
+		if err != nil {
+			return nil, errors.Wrap(err, "reading '='")
+		}
+		if equals != '=' {
+			return nil, errors.New("expected '='")
+		}
+		// NOTE: In the future, this may need to be parseItem() to
+		// support other Item types:
+		// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-4.2.5
+		value, err := parseString(reader)
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing value for param %s", name)
+		}
+
+		params[name] = value
+	}
+	return &parameterisedIdentifier{primaryID, params}, nil
+}
+
+// https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-07#section-4.2.7
+func parseString(reader *strings.Reader) (string, error) {
+	quote, err := reader.ReadByte()
+	if err != nil {
+		return "", errors.Wrap(err, "reading '\"'")
+	}
+	if quote != '"' {
+		return "", errors.New("expected '\"'")
+	}
+	var value strings.Builder
+	for {
+		char, err := reader.ReadByte()
+		if err != nil {
+			return "", errors.Wrap(err, "reading char")
+		}
+		if char <= 0x1f || char == 0x7f {
+			return "", errors.Errorf("invalid char %d", char)
+		}
+		if char == '"' {
+			break
+		}
+		if char == '\\' {
```

Comment:
shouldn't this be at line 161? according to 4.2.7 process point 4.2 happens before 4.4.

Reply:
Good catch. Yeah, I took a shortcut here moving the ifs around so that I could reuse the `value.WriteByte()` call below. It was (I think) equivalent behavior, but clearly not worth the confusion. Now it should read closer to the spec.

(Still a little different in the way it handles premature end-of-string: error-checking on the `ReadByte` instead of a condition on the `for` loop.)